src/opt/Optimizer/Optimizer.hs

```diff
@@ -128,3 +129,28 @@ dropBorrowBlocks = extend dropBorrowBlock
            ,decls = [([VarNoType name], target)]
            ,body}
       dropBorrowBlock e = e
+
+forwardExistingFuture = extend forwardExistingFuture'
+  where
+    forwardExistingFuture' e@(Forward{emeta, forwardExpr=arg@VarAccess{}})=Forward{emeta=emeta', forwardExpr=newExpr}
```

Comment:
I have added to more checks. `Forward{MessageSend}` and `Forward{FutureChain}` will be checked first and not be transformed. More test cases have been added.